File: server-py/app/scripts/ingest.py
"""
文档入库脚本，执行一次即可，知识库更新时重新执行
运行：python -m app.scripts.ingest
"""
import os

from langchain_core.documents import Document
from langchain_postgres import PGVector
from langchain_text_splitters import RecursiveCharacterTextSplitter

from app.db.postgres import PG_CONNECTION_STRING
from app.models.embedding import embeddings

COLLECTION_NAME = "knowledge_embeddings"

_KNOWLEDGE_DIR = os.path.join(os.path.dirname(__file__), "..", "data", "knowledge")


def _load_docs():
    files = ["products.md", "policies.md"]
    docs = []
    for file in files:
        with open(os.path.join(_KNOWLEDGE_DIR, file), "r", encoding="utf-8") as f:
            content = f.read()
        docs.append(Document(page_content=content, metadata={"source": file}))
    return docs


def ingest():
    print("开始处理文档...")

    docs = _load_docs()
    splitter = RecursiveCharacterTextSplitter(chunk_size=500, chunk_overlap=50)
    chunks = splitter.split_documents(docs)
    print(f"切分完成，共 {len(chunks)} 个片段")

    # 清空旧数据（全量更新场景）
    vector_store = PGVector(
        embeddings=embeddings,
        collection_name=COLLECTION_NAME,
        connection=PG_CONNECTION_STRING,
        use_jsonb=True,
        pre_delete_collection=True,
    )
    vector_store.add_documents(chunks)

    print("入库完成")


if __name__ == "__main__":
    ingest()

File: server/src/scripts/ingest.js
// server/src/scripts/ingest.js
// 文档入库脚本，执行一次即可，知识库更新时重新执行
// 运行：node src/scripts/ingest.js

import { readFileSync }        from 'fs';
import { join, dirname }       from 'path';
import { fileURLToPath }       from 'url';
import { RecursiveCharacterTextSplitter } from 'langchain/text_splitter';
import { PGVectorStore }       from '@langchain/community/vectorstores/pgvector';
import { Document }            from '@langchain/core/documents';
import { embeddings }          from '../models/embedding.js';
import { pool }                from '../db/postgres.js';

const __dirname = dirname(fileURLToPath(import.meta.url));

const PG_CONFIG = {
  pool,
  tableName: 'knowledge_embeddings',
  columns: {
    idColumnName:       'id',
    vectorColumnName:   'embedding',
    contentColumnName:  'content',
    metadataColumnName: 'metadata',
  },
};

const loadDocs = () => {
  const files = ['products.md', 'policies.md'];
  return files.map((file) => {
    const content = readFileSync(
      join(__dirname, '../data/knowledge', file),
      'utf-8'
    );
    return new Document({ pageContent: content, metadata: { source: file } });
  });
};

const splitter = new RecursiveCharacterTextSplitter({
  chunkSize:    500,
  chunkOverlap: 50,
});

const ingest = async () => {
  console.log('开始处理文档...');

  const docs   = loadDocs();
  const chunks = await splitter.splitDocuments(docs);
  console.log(`切分完成，共 ${chunks.length} 个片段`);

  // 清空旧数据（全量更新场景）
  const client = await pool.connect();
  try {
    await client.query(
      `CREATE TABLE IF NOT EXISTS knowledge_embeddings (
        id       bigserial PRIMARY KEY,
        content  text,
        metadata jsonb,
        embedding vector(1536)
      );`
    );
    await client.query('TRUNCATE knowledge_embeddings;');
  } finally {
    client.release();
  }

  await PGVectorStore.fromDocuments(chunks, embeddings, PG_CONFIG);

  console.log('入库完成');
  await pool.end();
};

ingest().catch((err) => {
  console.error('入库失败：', err.message);
  process.exit(1);
});
